src/targets/mobile/lib/withPersistentState.js

```diff
@@ -0,0 +1,20 @@
+import localforage from 'localforage'
+
+// The withPersistentState HOC automatically saves the wrappped components state and restores it when the component is mounted again.
+// Since the saved state will not be available immediately when the wrapped component is mounted, a `componentStateRestored` hook will be called once the state is restored
+export const withPersistentState = (WrappedComponent, persistenceKey) => class extends WrappedComponent {
+  async componentWillMount () {
+    if (super.componentWillMount) super.componentWillMount()
+    const persistedState = await localforage.getItem(persistenceKey)
+    this.setState(prevState => (persistedState || prevState), () => {
+      if (super.componentStateRestored) super.componentStateRestored()
+    })
```

Comment:
This is a valid concern and I actually tried this at first.

> If I understand well, you suggest to implement such a function to let the component do something once the state is set.

Specifically once it's set by the state restoration, yes. The use case here being that we want to increment a counter once the state has been restored, and only then.

While you *can* detect the state restoration in the wrapped component using `componentWillUpdate` or `componentDidUpdate`, it requires adding code to detect that the changes are coming from the restoration, and not a `setState` somewhere else. It also makes the whole code quite hairy because you end up with multiple `componentWillUpdate` hooks that feed into each other, to a point where it wasn't working at all. 

I'm sure it's possible, but:
- the extra hook provides a clean and easy way to achieve what we want
- it's completely optional

So yes, I think this is the best approach.